### PR TITLE
Add Gradebook APIs to Master

### PR DESCRIPTION
**Pull request by ShaydeNofziger**
# Description
This pull request adds a bunch of API functionality for the Grading API team. Additionally, the Grading team requests two changes to the `general` and `enrollment` APIs which are included in this PR.
## Requested Changes:
* `general`
  * `getStudent` We require data from the User model when retrieving a student. This shouldn't affect existing functionality for other consumers of the API. 
* `enrollment`
  * `getAllStudentSections` We require the ability to retrieve all student sections for a given section id. This is an entirely new API method, so other consumers shouldn't be affected.

## Known Issues
* Grade API is still a work in progress. As such, GPA is not being re-calculated / updated for a student. That is forthcoming.
* the User data returned with `getStudent` contains a hashed password. Ideally, that wouldn't happen. The fix would be to separate the `password` value from the `user` table and put it in a join table of sorts. But ehhh...who wants to code securely anyways?

### CODE REVIEW

**Comment by ShaydeNofziger** [comment]
@ZacharyMoran this PR includes requests for an addition to the enrollment API. Please review when you have a chance

File: API/Enrollment.php [hunk 1]
```diff
@@ -1287,6 +1297,38 @@ function withdrawFromWaitlist($studentID, $sectionID)
 	}
 }
 
+// retrieve all StudentSections for a given sectionID
+function getAllStudentSections($sectionID)
```
Comment (ZacharyMoran) [inline API/Enrollment.php]: This is literally the same function as getSectionEnrolled, the reason it is named as such is because there are two functions related to sections and students: getSectionEnrolled and getSectionWaitlist. The former is exactly what you have made here

File: API/API.php [hunk 2]
```diff
@@ -469,16 +469,16 @@ function getStudent($studentID)
 		$sqlite->enableExceptions(true);
 		
 		//prepare query to protect from sql injection
-		$query = $sqlite->prepare("SELECT * FROM Student WHERE USER_ID=:user_ID");
+		$query = $sqlite->prepare("SELECT * FROM Student as s INNER JOIN User as u ON s.USER_ID = u.ID WHERE s.USER_ID=:user_ID");
```
Comment (ZacharyMoran) [inline API/API.php]: This is covered by my group's "getStudentUser" function that does this functionality. This change is unnecessary.

File: API/Enrollment.php (hunk 1 above)
Comment (ShaydeNofziger) [inline API/Enrollment.php]: @ZacharyMoran My only issue with that function though is that we require the entire contents of the `Student_Section` model, not just the student_id.

Comment (ShaydeNofziger) [inline API/Enrollment.php]: So I either request that that function be updated to return all columns or we'll need this new function.

Comment (ZacharyMoran) [inline API/Enrollment.php]: I'll meet with my team tonight (at 5:30) and make sure changing the return on the other function wont break things, if it won't then that is the solution I prefer because it is nearly the same function repeated otherwise. I'll let you know

Comment (ShaydeNofziger) [inline API/Enrollment.php]: I completely agree. Thanks! I don't think it'll affect y'all, because I assume you're iterating the list and getting the value of `STUDENT_ID` prop.

Let me know and I can make the requested changes.

File: API/API.php (hunk 2 above)
Comment (ShaydeNofziger) [inline API/API.php]: Thank you! I wrote that a while back before merging your stuff. Will fix.

File: Database/createDB.bash [hunk 3]
```diff
@@ -18,4 +18,4 @@ then
 	sqlite3 $fileName < "TeamGrading.sql";
 fi
 
-chmod a+rwx $fileName;
\ No newline at end of file
+chmod a+rwx $fileName;
```
Comment (roda0313) [inline Database/createDB.bash]: Are you sure this works? I remember having issues where a bash script interprets new lines as a command, so i removed any new line characters. I really don't think this file should be changed at all. Check it out from the master branch and you should be good

File: Database/TeamGrading.sql [hunk 4]
```diff
@@ -3,30 +3,47 @@
...
 	STUDENT_SECTION_ID INTEGER NOT NULL,
-	FOREIGN KEY(STUDENT_SECTION_ID) REFERENCES StudentSection(ID)
-);
\ No newline at end of file
+	FOREIGN KEY(STUDENT_SECTION_ID) REFERENCES Student_Section(ID)
+);
+CREATE TRIGGER notification_created_time AFTER INSERT ON Notification
+BEGIN
+	UPDATE Notification SET CREATED_TIME = Datetime('now') WHERE ROWID = NEW.ROWID;
+END;
```
Comment (roda0313) [inline Database/TeamGrading.sql]: Since the database has not been changed in a while you may want to check that other teams don't have valuable data in the database since this will require a fresh DB for the changes to take effect

File: Database/DummyData.sql [hunk 5]
```diff
@@ -153,3 +154,5 @@ Grading Data
 INSERT INTO Grade (VALUE, IS_LOCKED, LAST_MODIFIED, STUDENT_SECTION_ID) VALUES (90, 0, "2017-03-28 15:30:00.000", 1);
 INSERT INTO Grade (VALUE, IS_LOCKED, LAST_MODIFIED, STUDENT_SECTION_ID) VALUES (75, 0, "2017-03-28 15:30:00.000", 2);
 
+INSERT INTO Comment (CREATED_TIME, USER_ID, GRADE_ID, CONTENT) VALUES ("2017-03-28 15:30:00.000", 1, 1, "I can't believe you didn't give me an A+!");
+INSERT INTO Comment (CREATED_TIME, USER_ID, GRADE_ID, CONTENT) VALUES ("2017-03-28 15:30:00.000", 2, 2, "I can't believe you didn't give me an A+!");
```
Comment (roda0313) [inline Database/DummyData.sql]: If you have triggers set up correctly the ``CREATED_TIME`` does not need to be input. Looking at your changes to triggers, you shouldn't need to specify a creation time

File: Database/TeamGrading.sql (hunk 4 above)
Comment (ShaydeNofziger) [inline Database/TeamGrading.sql]: Well yeah...but we're not doing database migrations so this is the only way to do it. No one should have "valuable data" in development.

File: Database/createDB.bash (hunk 3 above)
Comment (ShaydeNofziger) [inline Database/createDB.bash]: Did you experience that issue running the bash file or the batch file? POSIX standards define a line as terminating with a newline character. Without the newline, that command won't be run when executing the script on a unix system.

File: Database/DummyData.sql (hunk 5 above)
Comment (ShaydeNofziger) [inline Database/DummyData.sql]: Good catch. Will fix

Comment (ShaydeNofziger) [inline Database/DummyData.sql]: Resolved.

File: Database/createDB.bash (hunk 3 above)
Comment (ShaydeNofziger) [inline Database/createDB.bash]: And I just confirmed that the command does not get executed if there's no newline at EOF. I'm not going to revert this change.

Comment (roda0313) [inline Database/createDB.bash]: All i know is that on the VM hosting the API its working as is and that when I was creating the files (on a windows system) it adds a carriage return and a new line by default which caused issues early on

File: API/API.php (hunk 2 above)
Comment (ShaydeNofziger) [inline API/API.php]: I'm getting an error when querying the route you mentioned.

Here's the request:
`http://localhost:3333/API/API.php?userID=1&team=enrollment&function=getStudentUser`

Here's the response:
`An error has occurred`

Are you able to run that query correctly?

File: Database/createDB.bash (hunk 3 above)
Comment (ShaydeNofziger) [inline Database/createDB.bash]: But on windows, you should be running the batch file, not a bash file. This is how that should be written. I'm guessing either someone changed permissions on that DB manually, or that command is not necessary on the VM. Either way, this change is required for that command to be executed.

**Comment by ShaydeNofziger** [comment]
Submitting new PR
